src/queue.jl

```diff
@@ -1,38 +1,119 @@
 # Utilities for working with HSA queues
 
-"""
-An HSA queue.
-Each queue is uniquely associated with an agent.
-"""
-mutable struct HSAQueue
-    agent::HSAAgent
-    queue::Ref{Ptr{HSA.Queue}}
-    active::Bool
+@static if VERSION >= v"1.7-"
+    """
+    An HSA queue.
+    Each queue is uniquely associated with an agent.
+    """
+    @eval mutable struct HSAQueue
+        agent::HSAAgent
+        queue::Ptr{HSA.Queue}
+        @atomic active::Bool
+    end
+else
+    """
+    An HSA queue.
+    Each queue is uniquely associated with an agent.
+    """
+    mutable struct HSAQueue
+        agent::HSAAgent
+        queue::Ptr{HSA.Queue}
+        active::Bool
+    end
+    @inline function replacefield!(value, name::Symbol, expected, desired)
+        y = getfield(value, name)
+        ok = y === expected
+        if ok
+            setfield!(value, name, desired)
+        end
+        return (; old=y, success=ok)
+    end
 end
-
+const QUEUES = Dict{Ptr{HSA.Queue},WeakRef}()
+const QUEUE_STATUS = Dict{Ptr{HSA.Queue},HSA.Status}()
+const QUEUE_STATUS_COND = Dict{Ptr{HSA.Queue},Base.AsyncCondition}()
 const DEFAULT_QUEUES = IdDict{HSAAgent,HSAQueue}()
 
+async_send(data::Ptr{Cvoid}) = ccall(:uv_async_send, Cint, (Ptr{Cvoid},), data)
+
+function queue_error_handler(status::HSA.Status, _queue::Ptr{HSA.Queue}, data::Ptr{Cvoid})::Nothing
+    if status != HSA.STATUS_SUCCESS
+        l = RT_LOCK.cond_wait.lock
+        while true
```

Comment:
Instead of having a lock, how about each queue has a field for the associated condition? Or we pass the handle in as the data?

Reply:
That sounds excellent to me! Will do that with my next push.

Reply:
I also need to set the error status code, so I'll just pass in the pointer to the `HSAQueue` object itself.